src/inventory_ui.cpp

```diff
@@ -108,308 +393,413 @@ bool inventory_column::is_selected( const inventory_entry &entry ) const
 bool inventory_column::is_selected_by_category( const inventory_entry &entry ) const
 {
     return entry.is_item() && mode == navigation_mode::CATEGORY
-                           && entry.get_category_ptr() == get_selected().get_category_ptr()
-                           && page_of( entry ) == page_index();
+           && entry.get_category_ptr() == get_selected().get_category_ptr()
+           && page_of( entry ) == page_index();
 }
 
-const inventory_entry &inventory_column::get_selected() const {
+const inventory_entry &inventory_column::get_selected() const
+{
     if( selected_index >= entries.size() ) {
         static const inventory_entry dummy;
         return dummy;
     }
     return entries[selected_index];
 }
 
-std::vector<inventory_entry *> inventory_column::get_all_selected() const
+std::vector<inventory_entry *> inventory_column::get_all_selected()
 {
     std::vector<inventory_entry *> res;
 
     if( allows_selecting() ) {
-        for( const auto &entry : entries ) {
+        for( auto &entry : entries ) {
             if( is_selected( entry ) ) {
-                res.push_back( const_cast<inventory_entry *>( &entry ) );
+                res.push_back( &entry );
             }
         }
     }
 
     return res;
 }
 
-void inventory_column::on_action( const std::string &action )
+void inventory_column::on_input( const inventory_input &input )
 {
     if( empty() || !active ) {
         return; // ignore
     }
 
-    const auto move_selection = [ this ]( int step ) {
-        const auto get_incremented = [ this ]( size_t index, int step ) -> size_t {
-            return ( index + step + entries.size() ) % entries.size();
-        };
-
-        size_t index = get_incremented( selected_index, step );
-        while( entries[index] != get_selected() && ( !entries[index].is_item() || is_selected_by_category( entries[index] ) ) ) {
-            index = get_incremented( index, ( step > 0 ? 1 : -1 ) );
-        }
-
-        select( index );
-    };
-
-    if( action == "DOWN" ) {
+    if( input.action == "DOWN" ) {
         move_selection( 1 );
-    } else if( action == "UP" ) {
+    } else if( input.action == "UP" ) {
         move_selection( -1 );
-    } else if( action == "NEXT_TAB" ) {
+    } else if( input.action == "NEXT_TAB" ) {
```

Comment:
It's a dedicated [struct](https://github.com/CleverRaven/Cataclysm-DDA/pull/17765/files#diff-12ac5df926f33d8b6e6851de832dcddeR29).